Add Merlin's Typing recovery

xvw opened the pull request:
This PR upstreams Merlin's typing recovery to the OCaml Typechecker.

This new mechanism can be enabled with the `-typing-recovery` cli flag. When running in that mode, the typer won't stop at the first error. Instead it will continue typing as much as possible, reporting other errors along the way. 

## Why is this a useful patch from a Merlin maintenance perspective?

This is a major part of Merlin patches to the compiler and it enables reporting all errors in a file at once to the user's text editor. Additionnaly, having the option to not stop at the first error should be valuable to users that don't rely on Merlin when writing OCaml code.

Upstreaming these patches to the compiler will greatly simplify the process of upgrading Merlin to newer versions of the compiler, and we hope that it should also help maintaining a high-quality recovery by systematically handling new cases when they are introduced.

## About the design of recovery

The branch introduces a new module, `typing/typing_recovery.ml`, which provides a mechanism for raising exceptions without interruption: `raise_error`. It also introduces a new `ClFlags` (`--typing-recovery`), which enables typing recovery when passed.

The main idea behind recovery is to capture relevant type errors (this is essentially the part that will require caution when making changes to the type system, such as labelled tuples, for example). Instead of raising an exception, if the flag is present, the exception is stored (the mechanism is quite similar to the use of effects, where classic type checking would consist of discontinuing the continuation from the handler, and type checking with recovery would consist of continuing the programme), and when possible (notably in `typemod`), a placeholder value is constructed to leave a valid value. Errors can be recovered using `Typing_recovery.catch_errors`.

In addition to collecting errors, recovery maintains a register of transient types in error so as not to unnecessarily duplicate/propagate type errors (using `Typing_recovery.erroneous_type_register`).

#### Implication

Type error additions should be tested, in addition, with the `typing-recovery` flag to ensure that `raise_error` is used rather than `raise` in case of recovery.

## About this PR

If the implementation is not sufficiently clear, clarifications can be provided.

This PR is still a work-in-progress, but we think that it can provide a good basis to discuss how the remaining work should be done.

Among the subjects that need improvements there are:

- Proper reporting of errors in CLI mode. In the current state we print all the aggregated errors at the end of typing, before raising one last "Recovery happened" expection. One better option would be to print these errors at the moment they are reported, in a way similar to warnings. We are also not sure of which printing functions are best suited for that job (the primary objective being to be used by Merlin).
- There may be cases that are not recovered yet but could be.
- One challenge of introducing the recovery is to make sure new contributions take it into account (and we will be more than happy to help monitor the development of these new features and ensure that recovery is properly supported in these scenarios). How can we do that ?


I am available for synchronous reviews if this would facilitate the reviewers' work, and if the patch is approved, we will be available to the maintenance team to assist with the review of handling typing recovery in the future.

I did not introduce the typed holes feature that can helps recovery in Merlin, in order to try to keep the diff as small as possible.

## Further works

We are currently mining additional recovery cases in the Merlin code
base (which we will upstream in this PR or in the future depending on
its status).

I will also take into account any errors found by `check-typo` once the PR is sufficiently advanced in the review process (to avoid unnecessary _back-and-forth_).

xvw committed:
Add Typing Recovery

A mechanism that allows the type checker to detect more than just the
first type error, which is particularly useful for editor-side
instrumentation.

xvw committed:
Add `typing-recovery` before `modular explicit`

nojb commented:
I haven't looked at the code, but am strongly in favour of this.

For future reference, can you briefly mention what other components are keeping Merlin from being an ordinary executable depending on `compiler-libs`?

xvw commented:
> I haven't looked at the code, but am strongly in favour of this.
> 
> For future reference, can you briefly mention what other components are keeping Merlin from being an ordinary executable depending on `compiler-libs`?

Thank you!
The two main divergences are:

- Parsing recovery (which is the 'parsing' equivalent of typing recovery), even though it probably doesn't block the dependency on `compiler-libs`
- The way Merlin handles Short Paths which is completely different from the compiler's implementation

Having short-paths and typing-recovery (and probably some other small things like typed holes and utils) upstreamed would already allow us to rely on the compiler-libs typer instead of vendoring.

goldfirere commented:
🎉 Very excited for this!

trefis commented on Changes:
```diff
@@ -261,6 +261,13 @@ Working version
+- #14241: Introduction of a typing recovery mechanism to collect more
+  than one typing error, facilitating integration with editors that
+  need to report more than one error. Simplifying Merlin maintenance,
+  new versions of the compiler in new versions of the compiler.
```
That last sentence doesn't make sense.

xvw replied:
Removed!

trefis commented on testsuite/tests/typing-recovery/missing_constructor_in_function_norec.ml:
```diff
@@ -0,0 +1,10 @@
+   compile_only = "true";  
+   ocamlc.byte;
+   check-ocamlc.byte-output;
+*)
```
It's nice for the PR to highlight the differences with and without recovery, but apart from that I don't think this particular test is useful.
I suggest removing it.

xvw replied:
Yes, I try to keep `norec` tests only when it "seems" useful.

trefis commented on typing/ctype.mli:
```diff
@@ -80,7 +80,21 @@ val reset_global_level: unit -> unit
 val increase_global_level: unit -> int
 val restore_global_level: int -> unit
-        (* This pair of functions is only used in Typetexp *)
+(* This pair of functions is only used in Typetexp *)
```
nitpick: this whitespace diff is noise.

xvw replied:
Removed.

trefis commented on typing/ctype.mli:
```diff
@@ -80,7 +80,21 @@ val reset_global_level: unit -> unit
+val save_levels: unit -> levels
+
+(* Typing recovery specific. *)
+val set_levels: levels -> unit
```
IIUC these functions are useful when some "recoverable" type error occurs: in such situations, we are going to drop a subtree (e.g. an expression, or a pattern), synthesize a "dummy" node in its place (e.g. `_ : 'a`) and then resume typing the rest of the file.
However we don't know when precisely the error happened, and there might be an arbitrary number of `begin_def`s in need of closing before we can move on.
The `save_levels` / `set_levels` dance takes care of that.

However, explicit `begin_def` and `end_def` were retired a couple of years ago and the new `Ctype.with_[local_]level` takes care of cleaning up even in case of errors/exceptions.
So my question is : is the dance still necessary?

---
Note: the "global level" is still accessed through an "old style" API, but the only user (`Typetexp`) seems to have its own internal "new style" API.

xvw replied:
Yes, you are right! It is now removed (and makes the code easier). Thanks!

trefis commented on typing/typedecl.ml:
```diff
@@ -1227,7 +1227,13 @@ let transl_type_decl env rec_flag sdecl_list =
+           && (not (Typing_recovery.erroneous_type_check ty))
+           then raise(err) (* Strange but if typing recovery is enabled,
+                              it should only if erroneous_type_check. *)
+           else raise(err)
```
1. `if bool then foo else foo`? :thinking: 
2. I don't understand the comment
3. I went and had a look at what the code looks like in merlin, in case there's a difference, and there is, the code there is the following:
    ```ocaml
       match Ctype.closed_type_decl decl with
         Some ty ->
          if not (Msupport.erroneous_type_check ty) then
            raise(Error(sdecl.ptype_loc, Unbound_type_var(ty,decl)))
    ```
   That is: if `ty` is known to already be related to an error, do not complain about it.
   More precisely in this case: if `ty` has been synthesized (e.g. because `Typetexp` failed to "translate" the manifest) do not complain that it is unbound.

xvw replied:
Yes, I had a small confustion. Should be fixed now.

trefis replied:
It is fixed (although git's log shows me it wasn't when you posted! 😛), but I'm not fond of the style... let me try to use github's fancy features ... ⚙️

trefis commented on typing/typing_recovery.ml:
```diff
@@ -0,0 +1,143 @@
+  let saved_types = Cmt_format.get_saved_types () in
+  Cmt_format.set_saved_types [];
+  try
+    let result = with_warning_attribute ?warning_attribute f in
```
I don't like that part, `with_saved_types` already embeds a closure, why does it need to wrap around it?
Plus, it's not used consistently (there's a case in typemod where the attributes are no passed to `with_saved_types` and the closure does the call to `Builtin_attributes.warning_scope` itself).

xvw replied:
Is that better now?

trefis commented on typing/typecore.ml:
```diff
@@ -6217,6 +6572,10 @@ and map_half_typed_cases
     category env ty_arg ty_res loc caselist ~type_body ~check_if_total ->
+  let has_errors =
+    let monitor_errors = Typing_recovery.monitor_errors () in
+    not (!Clflags.typing_recovery) && !monitor_errors
```
The doc of `Typing_recovery.monitor_errors` states:
> Returns a reference initially set to false that will be set to true
    when a type error is raised.

But here you're directly reading that reference, before an error has a chance to occur.
This make the check a few lines below useless.

Funnily enough, this is caught by one of the tests of the PR : `type_recovery_test_1` (the tests need better name btw).

xvw replied:
Yes, fixed

trefis commented on typing/typecore.ml:
```diff
@@ -6421,11 +6780,13 @@ and map_half_typed_cases
-  else
-    (* Check for unused cases, do not delay because of gadts *)
-    unused_check false;
+  if not has_errors then begin
```
This check.

trefis commented on typing/typing_recovery.ml:
```diff
@@ -0,0 +1,143 @@
+
+let monitor_errors () =
+  let () = if !(!ref_monitor_errors)
+    then ref_monitor_errors := ref false
```
I believe a `bool ref ref` is surprising enough that it deserves an explanatory comment.

AFAICT the point here is to support nested calls to monitor errors where
- the "ancestor" call should always know if there was an error (either directly, or in one of the nested calls)
- a nested call should not be polluted by errors from the parent (or siblings)

xvw replied:
I attach your explanation as a commentary.

trefis replied:
Nitpicking territory here but:
1. I find the wording of your comment slightly confusing:
    > Using a reference handle allows nested calls to be monitored.
    
    Which nested calls? What does monitored mean?
2. the style / indentation ... brrr. I particularly dislike the `if` on the same line as the `let () =` that preceeds it, but beyond that: could we just have `if cond then assign;` ?

trefis commented on testsuite/tests/typing-recovery/type_recovery_test.compilers.reference:
```diff
@@ -0,0 +1,12 @@
+File "_none_", line 1:
```
That is not ok :)

xvw replied:
To be honest, I still do not know how to report properly multiples errors ? (Maybe @Octachron, do you have an idea to make it properly without making the diff bigger?)

trefis commented on testsuite/tests/typing-recovery/typing_recovery_unbound_type_constructor.compilers.reference:
```diff
@@ -0,0 +1,12 @@
+12 | let foo2 : int * toto = 3, 4
+                      ^^^^
+Error: Unbound type constructor toto
+
```
The error is duplicated / reported twice here.
Why is that?

xvw replied:
I try to figure why :/ it is very strange.

xvw replied:
Fixed!

trefis commented on testsuite/tests/typing-recovery/type_recovery_test_4.compilers.reference:
```diff
@@ -0,0 +1,6 @@
+       11 | and (a,b) = (1,2)
+                ^^^^^
+       Error: Only variables are allowed as left-hand side of let rec
+
```
I feel like depending on the way in which the error is printed to know whether the recovery happened is rather fragile.

Adding a second error later in the test would confirm that we indeed continued after the first one.
Furthermore, if we want to offer stronger guarantees regarding the actual recovery, we could try and make use of the recovered parts as well; and confirm we actually exist / have the type we expect / do not cause a cascade of errors when used / etc.

trefis commented on typing/typing_recovery.mli:
```diff
@@ -0,0 +1,51 @@
+val recovery_attributes : Parsetree.attributes -> Parsetree.attributes
+
+(** Turn saved types from Cmt_format into attributes *)
+val flush_saved_types : unit -> Parsetree.attributes
```
This is already called from `recovery_attributes`, does it need to be exposed? Or, rather: is it OK to call this function instead of `recovery_attributes`?

xvw replied:
Yes, removed from the public API.

trefis commented on typing/typecore.ml:
```diff
@@ -2229,7 +2383,25 @@ and type_pat_aux
+              pat_extra = [extra];
+            }
+        | _, p ->
+            { p with pat_type = ty; pat_extra = extra::p.pat_extra }
```
This is reintroducing code that was removed in #13806.

xvw replied:
Well spotted ;) thanks!

trefis commented on typing/typing_recovery.mli:
```diff
@@ -0,0 +1,51 @@
+val catch_errors : Warnings.state -> exn list ref -> (unit -> 'a) -> 'a
+
+(** Re-raise an exception*)
+val resume_raise : exn -> 'a
```
The docstring is incorrect.
I went and had a look in merlin, which states:
```ocaml

(** Resume after error: like [raise_error], but if a handler was provided a
    Resume exception is raised.  This allows to specify a special case when an
    error is caught. *)
exception Resume

val resume_raise : exn -> 'a
```

Also, `resume_raise` is only used in a single place in typecore, and I feel like it would be cleaner to special case that occurrence and not bother with `resume_raise`.

xvw replied:
I try to _inlining it_ directly inside typecore without intermediate exception, is that ok for your?

trefis commented on typing/typing_recovery.ml:
```diff
@@ -0,0 +1,143 @@
+      | Ctype.Unify _
+      | Failure _ ->
+          (* TODO: Should log an error *)
+          ()
```
This looks fishy, and not just because of the TODO comment.
Can you explain why there are these special cases?

trefis commented on typing/typing_recovery.ml:
```diff
@@ -0,0 +1,143 @@
+  with exn ->
+    let saved_types' = Cmt_format.get_saved_types () in
+    Cmt_format.set_saved_types (saved_types' @ saved_types);
+    reraise exn
```
I don't completely understand this code.

When an exception is raised by `f` (which I understand as "recovery didn't happen") it makes sure to add whatever types `f` "saved" to whatever types where already saved prior to the call.
That seems reasonable (although if the saved types weren't cleared before the call there would be no need).

When `f` doesn't raise, that is: either there was no error or it was recoverable, then it adds to the saved types when called with `save_part`... and in the other case just leaves the list in the state that `f` put it in (so `saved_types`, from our snapshot, are dropped).

Now, if you look at `Typecore.type_expect`, in the "normal" path (without recovery), the expression is "saved" when typing succeeds, but in the recovery paths `save_part` is not passed, and the expression is not saved (and whatever was already saved is dropped).

Again, I'm not sure that the behavior I described here is accurate, but that's the point.
I'm not sure what happens here, but I also don't understand the intent behind this code.

---

I have some idea why all this machinery is in place:
- the saved types are attached to recovered nodes using attributes (this is obvious from the PR)
- ... so that merlin can have access to nodes "under" type errors to do whatever it needs doing. This is not obvious from the PR: it is not explained nor mentioned anywhere. It should be.

But even knowing that, I feel like the code remains obscure.

trefis replied:
Note (to self): my ramblings above have made their way into the code as comments, but as I wasn't able to really explain anything when I posted them, I am still unhappy with that function.

I'm hopeful this will get resolved with @xvw's ongoing extensive testing/exploration of the recovery code.

trefis reviewed (changes requested):
I've given the PR a first pass, and have left a number of comments that I'd like to see addressed. I'll move on to a deeper review once those are resolved.

In its current state the PR:
- brings code that has been rotting in merlin for some years, providing a strong illustration of why it'd be better to have that code in the compiler.
- introduce regressions compared to merlin that are observable in the tests the PR adds.
- lacks a great deal of explanation / documentation; which I believe would be instrumental in making sure the recovery code evolves with the compiler

xvw commented:
Thanks a lot @trefis!
I'll get back to it at the end of the week.

trefis commented on typing/typedecl.ml:
```diff
@@ -1227,7 +1227,13 @@ let transl_type_decl env rec_flag sdecl_list =
+             if not (Typing_recovery.erroneous_type_check ty) then
+               raise(err)
+             else ()
+           else raise(err)
```
```suggestion
           if !Clflags.typing_recovery && Typing_recovery.erroneous_type_check ty then
               () (* do not report spurious error, we're in a recovered context *)
           else raise(err)
```

trefis commented on typing/typecore.ml:
```diff
@@ -2930,33 +3082,44 @@ let collect_unknown_apply_args env funct ty_fun0 rev_args sargs =
+                if !Clflags.typing_recovery then
+                  let () = raise_error exn in
+                  newvar (), ty_fun
+                else raise exn
```
Somewhat of a style comment, but I'm not fond of the `with exn` catch-all.
I would feel better if this bit was written like this:
```suggestion
              with Error _ as exn when !Clflags.typing_recovery ->
                raise_error exn;
                newvar (), ty_fun
```

trefis commented on typing/typing_recovery.ml:
```diff
@@ -0,0 +1,178 @@
+       from [f] are preserved. *)
+    let result = match warning_attribute with
+      | None -> f ()
+      | Some attr -> Builtin_attributes.warning_scope attr f
```
Since the code changed/disappeared github fails to show my previous comment in context, so I'm starting a new one here.

Basically I was complaining that `with_saved_types` was wrapping the closure it takes as an argument (I formulated vaguely with "adds one level of wrapping" or something of the sort).
In bf29436 you addressed that comment by inlining the local wrapper calling `Builtin_attributes` ... which is the call I'd actually like to see disappear from here: let the callers handle the warning scope !

At the places where `with_saved_types` is called there currently is another branch/code-path when recovery is not enabled where `Builtin_attributes...` is called. I think it should be explicit there in both branches.
"Hiding" it in `with_saved_types` introduces spurious differences between the code paths that we should try to avoid.

trefis commented on typing/typecore.ml:
```diff
@@ -4126,7 +4330,11 @@ and type_expect_
-                raise (Error (loc, env, err))
+                if !Clflags.typing_recovery then
+                  let () = raise_error(error (loc, env, err)) in
+                  ty_expected
```
This is not the right place to recover!

You're trying to recover from inside a check which doesn't "produce" anything, so you're just allowing the check to run further. And I thought this would result in spurious errors (especially since `expected_ty` is not at all a correct fallback here), but actually you get a compiler crash (fun !).

If I take the example you added to the testsuite but make it "longer" I get this:
```
❯ cat /tmp/test.ml
type (_, _) eq = Eq : ('a, 'a) eq

let bad : type a. ?opt:(a, int -> int -> int) eq -> unit -> a =
  fun ?opt:(Eq = assert false) () x y -> x + y

let x = 10
❯ ./runtime/ocamlrun ./ocamlc -c -typing-recovery -nostdlib -I stdlib /tmp/test.ml
>> Fatal error: Label_mismatch not expected as this point; this shouldhave been caught when the function was typechecked.
File "_none_", line 1:
Error: File "/tmp/test.ml", line 4, characters 2-46:
       4 |   fun ?opt:(Eq = assert false) () x y -> x + y
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       Error: The syntactic arity of the function doesn't match the type constraint:
              This function has 4 syntactic arguments, but its type is constrained to
                ?opt:(a, int -> int -> int) eq -> unit -> a.
               Hint: consider splitting the function definition into
                 fun ... gadt_pat -> fun ...
                 where gadt_pat is the pattern with the GADT constructor that
                 introduces the local type equation on a.
```

trefis commented on typing/typecore.ml:
It is ugly but a quick fix for the issue I pointed above would be to wrap this whole block in
```ocaml
try ...
with Error _ as exn when !recovery ->
  raise_error exn
```

trefis commented on typing/typing_recovery.ml:
```diff
@@ -0,0 +1,178 @@
+  | [] -> attrs'
+  | attrs -> attrs' @ attrs
+
+module Error_set = Set.Make (struct
```
The use of a set might be correct ... but I have an uneasy feeling about it.
For one thing it seems heavy weight (do we really have that many errors that it makes sense to build a set instead of checking for membership before inserting in a list?) but also I have this vague sensation that it is just trying to sweep an fundamental issue under the rug: we're trying to hide duplicated errors, when we shouldn't have duplicated errors.

I could very well be incorrect: perhaps duplicated errors are unavoidable. But I'd like to be convinced of that fact.
For instance: I pointed out duplicated errors in my previous review, this new code hides them, but could I get an explanation of why they were present in the first place?

trefis replied:
I'm learning to live with this state of the world, feel free to mark as resolved.

trefis reviewed (commented):
Here's another partial review of the PR, mostly looking over @xvw changes in answer to my earlier comments.

While I was doing this second round of review @xvw informed me that he was in the process of testing recovery more extensively, and fixing bugs as he was finding them. So I took a pause while he was doing that.
He finished a couple of days ago and indeed added a lot of tests (I'm really getting that warm Christmas feeling right now!) ... which I haven't looked at yet! 🥲 

Anyway, this PR is shaping nicely and I'll try to jump into the meat of things soon!

Until then, I have another (non-blocking) high-level style comment: this PR is sprinkling checks over the typechecker to see whether recovery is on or not, sometimes "high up" in the pipeline (in particular in typemod and typecore). In doing so, it is creating different code paths which IMO are undesirable: it's more maintenance work in the future, but also increasing the cognitive load on people trying to understand debug the typechecker.
I wish these recovery checks were pushed down as much as possible, so they can be avoided in many places (mostly they also already exist in the "leaves", so it's probably just a matter of removing redundant checks).
Note: I realise this comment is a bit hand-wavy, I'll likely have a more synchronous discussion with @xvw about it.

xvw committed:
Make things compiles

xvw committed:
Move argument handler

xvw committed:
Add missing case for mutual recursion

xvw committed:
Use type ascription instead of deconstruction outside of open Types

xvw committed:
Raise error on `Repeated_tuple_***_label`

xvw committed:
Add test for modular explicit in presence of recovery

trefis committed:
step 1: can't raise Typecore.Error from outside the Typecore.Error module

trefis committed:
deprecate Typing_recovery.raise_error

trefis committed:
switch Typecore to the new API
Also added some TODOs and FIXME

trefis committed:
propagate Typecore intf change to callers

trefis committed:
catch and log Env errors raised from typecore

trefis committed:
one less FIXME: uncatch errors in partial_pred

trefis committed:
shorter paths in tests/typing-recovery

trefis committed:
slightly more specific test

trefis committed:
promote testsuite:
- flag recovery issue caught by the testsuite
- there are a couple of problems that ought to disappear once the typer
  switches to the new API

trefis committed:
some doc

xvw committed:
Fix a small typo and do not use docstring

xvw committed:
Private recoverable errors in `typtexp`

xvw committed:
Expand the new API in `typemod`

xvw committed:
Addressing some comments in `typecore`

xvw committed:
Remove residual exception in `typetexp`

xvw committed:
Remove residual exception in `typemod`

xvw committed:
Fix toploop

xvw committed:
Remove already logged exception

xvw committed:
Remove already logged exception in `typemod`

xvw committed:
Fix shadowing for `typemod`

xvw committed:
Use fancy trick for private exception constructor

xvw committed:
Rename `Error.Error` into `Error.In_context`

xvw committed:
Use fancy trick for private exception constructor in `typemod`

xvw committed:
Use fancy trick for private exception constructor in `typecore`

xvw committed:
Change `ty_param`

xvw committed:
Less-precise tests

xvw committed:
Build set from right to left

xvw committed:
Handle `Error.In_context` for `Env`

xvw committed:
Loging test for sciences

xvw committed:
Split Typing_recovery and Typing_recovery_state

xvw committed:
Use `typing recovery` log

Also make depend to track new dependency

xvw committed:
Re-promote tests for observation

trefis committed:
remove FIXME: it's fixed

trefis committed:
env: don't log errors that are caught

trefis committed:
typetexp: recover lookup errors

xvw committed:
Move Typeclass to private error handling

xvw committed:
Fix Illegal Class Expr test

xvw committed:
User facing Error handling in `typedecl`

xvw committed:
User facing Error handling in `persistent_env`

xvw committed:
Remove `env` from error handling in `typedecl`

xvw committed:
Simplify errors for `Persistent_env`

xvw committed:
Addressing some fixme/todo

xvw committed:
Some fixes

trefis commented on testsuite/tests/typing-recovery/apply_non_function.ml:
```diff
@@ -0,0 +1,21 @@
+let z = 10
+let a = z + 10
+
+let e : string = 10
```
Seems like noise.
We've tested the same error twice already, so we know it's being recovered, there's no need to check that again.

trefis commented on testsuite/tests/typing-misc/typecore_errors.ml:
```diff
@@ -493,3 +493,15 @@ Error: Type "u -> unit" is not a subtype of "v -> unit"
+Error: This function expects too many arguments, it should have type
+       "'a -> int"
+|}]
+
```
This test is does not exercise recovery, and is redundant with tests in `.../typing-misc/apply_non_function.ml`
It should be removed.

xvw replied:
It is in `typing-misc`, @Octachron ask me to add the missing case in typecore error.

trefis commented on testsuite/tests/typing-recovery/apply_wrong_label.ml:
```diff
@@ -0,0 +1,24 @@
+
+let x = 10 
+
+let y = f ~x
```
This feels like noise as well: the code is valid and unused in the rest of the file.

trefis commented on testsuite/tests/typing-recovery/cannot_infer_signature.ml:
```diff
@@ -0,0 +1,34 @@
+let f =
+  let m = (module struct let x = 10 end) in
+  let n = (module struct let x = 10 end : S) in
+  m, n
```
`m` and `f` are redundant with `U.f`, and the recovery on that one is tested more extensively, so let's remove these two and keep just `U`?

trefis commented on testsuite/tests/typing-recovery/coercion_failure.ml:
```diff
@@ -0,0 +1,26 @@
+
+let f =
+  let x = ([`B] :> [`A]) in
+  `A :: x
```
Again, the same thing is checked several times.
This is both redundant and noisy.

trefis commented on testsuite/tests/typing-recovery/extra_missing_tuple_label.ml:
```diff
@@ -0,0 +1,32 @@
+
+let c =
+  let ~x, k1, ~y, ~z, ~w, k2, k3 = lt in
+  ()
```
That is just `a` renamed.

trefis commented on testsuite/tests/typing-recovery/function_arity_type_clash.ml:
```diff
@@ -0,0 +1,22 @@
+let x = 10
+
+let bad2 : type a. ?opt:(a, int -> int) eq -> unit -> a =
+  fun ?opt:(Eq = assert false) () x -> x + 1
```
That seems redundant/noisy.

trefis commented on testsuite/tests/typing-recovery/incoherent_label_order.ml:
```diff
@@ -0,0 +1,18 @@
+let u f x y =
+  match f ~x ~y, f ~y ~x with
+  | 10, x -> 20, x
+  | x, y -> x + 10, y
```
This is noise.

trefis commented on testsuite/tests/typing-recovery/inside_check.ml:
```diff
@@ -0,0 +1,15 @@
+type (_, _) eq = Eq : ('a, 'a) eq
+
+let bad : type a. ?opt:(a, int -> int -> int) eq -> unit -> a =
+  fun ?opt:(Eq = assert false) () x y -> x + y
```
What is this supposed to test?
It's the same code (and error) as in `function_arity_type_clash.ml`

xvw replied:
It was illustrating a compilation error (that you notice during one of the first batch of review)

trefis commented on testsuite/tests/typing-recovery/instance_variable_not_mutable_norec.ml:
```diff
@@ -0,0 +1,27 @@
+  in
+  o # foo
+
+let x : string = 11
```
let's remove this test, it's not testing anything that isn't already checked somewhere else.

trefis commented on testsuite/tests/typing-recovery/invalid_atomic_loc_payload.ml:
```diff
@@ -0,0 +1,21 @@
+module F = struct
+  type t = { mutable x : int [@atomic] }
+  let f t = [%atomic.loc t]
+end
```
Again, testing the same thing twice

trefis commented on testsuite/tests/typing-recovery/label_missing.compilers.reference:
```diff
@@ -0,0 +1,19 @@
+File "label_missing.ml", line 14, characters 8-32:
+14 | let _ = {x = 10; y = 10; y = 11}
+             ^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The record field label y is defined several times
```
Here we could have expected an error about `z` being undefined, but as it's sharing the same location as the one we have, it is not being displayed.

Which is probably not the end of the world ... it might even be a good thing. But perhaps add a comment in `label_missing.ml` to explain what's going on?

trefis commented on testsuite/tests/typing-recovery/missing_constructor_and_invalid_types_norec.ml:
```diff
@@ -0,0 +1,15 @@
+  match x with
+  | A -> ()
+  | B -> 3
+  | C -> 'a'
```
Let's remove this test. We can assume that without recovery, things are already well tested.

trefis commented on testsuite/tests/typing-recovery/missing_constructor_in_function.ml:
```diff
@@ -0,0 +1,11 @@
+*)
+
+type t = A | B
+let f (C : t) : int = ()
```
Nitpick, can we fold this test in the other "missing constructor" one?

trefis commented on testsuite/tests/typing-recovery/multiply_bound_variable.ml:
The filename should be `multi...` not `mulit...` :)

trefis commented on testsuite/tests/typing-recovery/no_value_clause.ml:
```diff
@@ -0,0 +1,30 @@
+      | exception Not_found -> 1
+      | exception Failure _ -> 2
+    in a + 10
+end
```
Checking the same thing a few too many times.
Probably drop `R` here? If the recovery happens at the level of expressions, there's no need to wrap in a module.

trefis commented on typing/env.ml:
```diff
@@ -672,12 +672,31 @@ type error =
 
-exception Error of error
+module Error : sig
+  type exn += private In_context of Location.t * t * error
```
Let's not require an environment if we don't need to.
Also the location is redundant: it's already included in every constructor of the `error` type.

trefis commented on typing/env.ml:
```diff
@@ -672,12 +672,31 @@ type error =
+    if !Clflags.typing_recovery then
+      Typing_recovery.log_and_raise err
+    else
+      raise err
```
Since, unlike what's happening in typecore, we're not doing any extra computation could this just be `Typing_recovery.log_and_raise`?

trefis commented on typing/env.ml:
```diff
@@ -1996,7 +2016,7 @@ and check_value_name name loc =
     for i = 1 to String.length name - 1 do
       if name.[i] = '#' then
-        error (Illegal_value_name(loc, name))
+        Error.log_and_raise loc empty (Illegal_value_name (loc, name))
```
I suspect this could be a `log_or_raise`... but there's no test for this error in presence of recovery.
Could you add one and then see if switching to `log_or_raise` makes a difference?

trefis commented on typing/typeclass.ml:
```diff
@@ -109,9 +109,30 @@ type error =
+      Typing_recovery.log_and_raise err
+    else
+      raise err
+
```
Same question/comment as in `Env`.

trefis commented on typing/typeclass.ml:
Could you add a test for this error in presence of recovery?
Considering how the `Error_set` works I think we should only see the `Illegal_reference_to_recursive_class_type` one (even though both will have been logged), but it'd be good to confirm, and prevent regressions.

trefis commented on typing/typecore.ml:
```diff
@@ -219,7 +219,143 @@ type error =
+      Ideally we'd do that by making the exception private, but that is
+      currently not possible, so we instead make its argument private. Thus
+      ensuring that these errors can only be reported using the functions
+      exposed here. *)
```
That's obviously wrong.

trefis commented on typing/typecore.ml:
```diff
@@ -1008,6 +1142,7 @@ let solve_constructor_annotation
           new_local_type ~loc:name.loc Definition
             ~manifest_and_scope:(tv, Ident.lowest_scope) in
         let (id, new_env) =
+          (* TODO: Ensure that long path is not ensured. *)
```
?

trefis commented on typing/typecore.ml:
```diff
@@ -1169,9 +1303,13 @@ let solve_Ppat_record_field loc penv label label_lid record_ty =
+      (* FIXME: in case of recovery the [Pattern_type_clash] will already have
+         been logged, so there will be two errors reported, but we'd want only
+         one.
+         ... add a [Typing_recovery.unlog : exn -> unit]? *)
```
This `FIXME` is incorrect.

trefis commented on typing/typecore.ml:
```diff
@@ -1761,10 +1902,20 @@ let split_half_typed_cases env zipped_cases =
+           then Error.log_or_raise ...;
+           add_case ... vp, add_case ... ep
+         ]}
+      *)
```
I briefly tried but couldn't easily construct a test case where that'd be true.
Given that merlin doesn't appear to have suffered of a lack of recovery in this part over its years of existence, let's just remove that TODO.

trefis commented on typing/typecore.ml:
```diff
@@ -1876,10 +2025,35 @@ let rec type_pat
+    if !Clflags.typing_recovery then
+      Typing_recovery_state.with_saved_types (fun () ->
+          try delayed ()
+          with Error.In_context _ ->
```
Here you should also catch `Env.Error.In_context` and `Typetexp.Error.In_context` (i.e. the two other kind of errors that can be raised when typing a pattern).

To observe a difference of behavior compare:
```ocaml
type i = I of int

let f = function
  | I "foo" ->
      3.14 + 2
```
which reports two errors, with:
```ocaml
let f = function
  | J ->
      3.14 + 2
```
which only reports one, or with
```ocaml
let f = function
  | `azdwbie
  | `c7diagq ->
      3.14 + 2
```

(And actually, for that last one, the error is logged and recoverable ... but currently never recovered from!)

trefis commented on typing/typecore.ml:
```diff
@@ -1876,10 +2025,35 @@ let rec type_pat
+               such as [Need_backtrack], etc.
+
+               N.B. the error has already been logged, we only need to recover
+               and mark the type so has to not generate a cascade or errors. *)
```
Actually, the "marking" (i.e. `erroneous_type_register`) is not used to prevent other errors from appearing, but to prevent warnings.

trefis commented on typing/typecore.ml:
```diff
@@ -2754,13 +2930,15 @@ let partial_pred ~lev ~splitting_mode ?(explode=0) env expected_ty p =
-    (* types are invalidated but we don't need them here *)
-    Some typed_p
-  with Error _ | Empty_branch ->
+    Typing_recovery.uncatch_errors (fun () ->
```
I think it's worth adding a comment here explaining why `uncatch_errors` is called.

trefis commented on typing/typecore.ml:
```diff
@@ -3527,7 +3717,10 @@ let type_approx_fun_one_param
-      raise (Error(loc_fun, env, err))
+      let level = get_level (instance ty_expected) in
+      Error.log_or_raise loc_fun env err;
+      { ty_param = newvar2 level; ty_ret = ty_expected}
```
The recovery is not correct here, `ty_param` should be a poly type.
On this input:
```ocaml
let rec id_wrong_ty : int = fun x -> x
```
I get:
```
$ ./runtime/ocamlrun ./ocamlc -c -i -nostdlib -I stdlib -typing-recovery /tmp/test.ml
Fatal error: exception File "typing/btype.ml", line 187, characters 9-15: Assertion failed
```

trefis commented on typing/typecore.ml:
```diff
@@ -4030,10 +4223,13 @@ let with_explanation explanation f =
+             (loc', env', Expr_type_clash(err', None, exp'))
         when not loc'.Location.loc_ghost ->
+          (* FIXME: as in solve_Ppat_record_field we're refining the error and
+             when recovery is on it ends up reported twice. *)
```
As the one it refers to, this FIXME is incorrect.

trefis commented on typing/typecore.ml:
```diff
@@ -4134,19 +4330,45 @@ let rec type_exp ?recarg env sexp =
+  if !Clflags.typing_recovery then
+    Typing_recovery_state.with_saved_types (fun () ->
+        try delayed ()
+        with (Error.In_context _ | Env.Error.In_context _) ->
```
As in `type_pat` you want to also catch `Typetexp.Error.In_context` otherwise on things like `let x =  `azdwbie = `c7diagq` you get no recovery.

trefis commented on typing/typecore.ml:
```diff
@@ -4575,147 +4797,168 @@ and type_expect_
-            fields; representation;
-            extended_expression = opt_exp
+      (* Recovery is done here instead of in [type_expect] as for the other
+         case, because being a bit more precise about the what is recovered (a
```
```suggestion
         cases, because being a bit more precise about what is recovered (a
```

trefis commented on typing/typecore.ml:
```diff
@@ -4575,147 +4797,168 @@ and type_expect_
+      (* Recovery is done here instead of in [type_expect] as for the other
+         case, because being a bit more precise about the what is recovered (a
+         record) allows various merlin queries to perform better.
+         For instance if try completion inside the {} it will propose the
```
```suggestion
         For instance if one tries completion inside the {} merlin will propose the
```

trefis commented on typing/typecore.ml:
```diff
@@ -5246,26 +5504,38 @@ and type_expect_
            exp_env = env }
 
   | Pexp_struct_item (si, e) ->
+      (* TODO: Maybe a missing opportunity for Typing Recovery here. *)
```
?

trefis commented on typing/typecore.ml:
```diff
@@ -5246,26 +5504,38 @@ and type_expect_
+
+           N.B. the original error will already have been logged. *)
+        Typing_recovery.catch_errors (Warnings.backup ()) (ref [])
+          (fun () -> type_expect env e ty_expected_explained)
```
The comment is slightly out of date (the structure item is not necessarily an open), although the logic is sound.
I'm a bit worried about the end of the comment though "the original error will already have been logged", that's only true if we're catching a "recoverable" error, but given that this catches any exception ...

I'd feel much better if you replaced the wildcard pattern by 
```ocaml
Typemod.Error.In_context _ | Error.In_context _ | Env.Error.In_context _ | ...
```
but typecore can't depend on typemod in that way.
So maybe something like `exn when !Clflags.typing_recovery and is_recoverable exn`

With somewhere a:
```ocaml
(* to be filled in by typemod *)
let is_typemod_recoverable_error : (exn -> bool) ref = ref (fun _ -> false)

let is_recoverableexn =
  match exn with
  |  Error.In_context _ | Env.Error.In_context _ | Typetexp.Error.In_context _ | ... -> true
  | _ -> !is_typemod_recoverable_error exn
```

And actually you'd probably want to use `is_recoverable` in `type_pat` and `type_expect` and other places like that.

trefis commented on typing/typecore.ml:
```diff
@@ -5504,7 +5774,10 @@ and split_function_ty env ty_expected ~arg_label ~has_poly ~first ~in_function =
           error_of_filter_arrow_failure ~explanation ty_fun err ~first
         in
-          raise (Error(loc, env, err))
+        Error.log_or_raise loc env err;
+        let level = get_level (instance ty_expected) in
+        let ty_param = newty2 ~level (Tpoly (newvar2 level, [])) in
+        { ty_param; ty_ret = ty_expected}
```
This is what the fix to https://github.com/ocaml/ocaml/pull/14241/changes#r2852413813 should look like.

trefis commented on typing/typecore.ml:
```diff
@@ -5580,8 +5853,14 @@ and type_function
+        try
+          with_explanation ty_fun.explanation (fun () ->
+              unify_exp_types loc env exp_type (instance ty_expected))
+        with _ -> Typing_recovery.erroneous_type_register ty_expected
```
😱 
```suggestion
        with exn when is_recoverable exn -> Typing_recovery.erroneous_type_register ty_expected
```

Also, `!Clflags.typing_recovery` could be part of the guard, so we don't have to have two branches.

trefis commented on typing/typecore.ml:
```diff
@@ -5664,8 +5943,16 @@ and type_function
+        if !Clflags.typing_recovery then
+          try suspended ()
+          with _ -> Typing_recovery.erroneous_type_register ty_expected
+        else suspended ()
```
As above.
And here we can then make the `suspended` wrapping disappear.

trefis commented on typing/typecore.ml:
```diff
@@ -5894,14 +6181,34 @@ and type_label_access env srecord usage lid =
-      (Label.disambiguate usage lid env expected_type) labels in
-  (record, label, expected_type)
+  try suspended ()
+  with (Env.Error.In_context _ | Error.In_context _) when !Clflags.typing_recovery ->
```
Likewise (regarding `is_recoverable` and the `suspended` wrapping)

trefis commented on typing/typecore.ml:
```diff
@@ -6319,6 +6626,38 @@ and type_argument ?explanation ?recarg env sarg ty_expected' ty_expected =
+  if !Clflags.typing_recovery then
+    Typing_recovery_state.with_saved_types (fun () ->
+        try delayed ()
+        with Error.In_context _ ->
```
Like `type_pat` and `type_expect`:
```suggestion
        with exn when is_recoverable exn ->
```

trefis commented on typing/typecore.ml:
```diff
@@ -6549,6 +6888,10 @@ and type_construct env ~sexp lid sarg ty_expected_explained =
 and type_statement ?explanation env sexp =
+  let recovery_errors = Typing_recovery.monitor_errors () in
+  let has_recovery_errors () =
+    (!Clflags.typing_recovery) && !recovery_errors
```
Is the `(!Clflags.typing_recovery)` part really needed?
If recovery is not enabled, an exception will have preempted the rest of the execution anyway, right?

trefis commented on typing/typecore.ml:
```diff
@@ -6615,6 +6959,10 @@ and map_half_typed_cases
     category env ty_arg ty_res loc caselist ~type_body ~check_if_total ->
+  let recovery_errors = Typing_recovery.monitor_errors () in
+  let has_recovery_errors () =
+    (!Clflags.typing_recovery) && !recovery_errors
```
Same remark.

trefis commented on typing/typecore.ml:
```diff
@@ -6911,8 +7261,16 @@ and type_function_cases_expect
+      | _ ->
+          let () = Typing_recovery.erroneous_type_register ty_expected in
+          cases, partial, ty_fun
+    else delayed ()
```
blabla `is_recoverable` blabla `delayed` wrapper blabla

🙂

trefis commented on typing/typecore.ml:
```diff
@@ -7078,6 +7436,12 @@ and type_let ?check ?check_strict
+      (fun {vb_pat=pat} -> match pat.pat_desc with
+           Tpat_var _ -> ()
+         | _ -> Error.log_or_raise pat.pat_loc env Illegal_letrec_pat)
+      l;
```
Seems like a rebase issue: that's introducing code that already exists ~100 lines above.

trefis commented on typing/typedecl.ml:
```diff
@@ -81,13 +81,35 @@ type error =
+    if !Clflags.typing_recovery then
+      Typing_recovery.log_and_raise err
+    else
+      raise err
```
And same question as in the other files again: could this just be `Typing_recovery.log_and_raise` in both branches?

trefis commented on typing/typedecl.ml:
```diff
@@ -1234,7 +1263,13 @@ let transl_type_decl env rec_flag sdecl_list =
-         Some ty -> raise(Error(sdecl.ptype_loc, Unbound_type_var(ty,decl)))
+         Some ty ->
+           if !Clflags.typing_recovery &&
+              Typing_recovery.erroneous_type_check ty then
```
`!Clflags.typing_recovery &&`  is not necessary, `erroneous_type_check` would return false anyway in the absence of recovery.

So this code can become:
```ocaml
if not (Typing_recovery.erroneous_type_check ty) then
  Error.log_and_raise ...
```
as it already is in merlin.

(The comment can stay in some form though, comments are nice)

trefis commented on typing/typedecl.ml:
```diff
@@ -1668,8 +1711,8 @@ let transl_value_decl env loc valdecl =
-  | [] ->
-      raise (Error(valdecl.pval_loc, Val_in_structure))
+    | [] ->
+        Error.log_and_raise valdecl.pval_loc Val_in_structure
```
I've try to mostly not report on these, but the indentation change here hurts me (because it makes the indentation of that case inconsistent with the others).

trefis commented on typing/typedecl_separability.ml:
```diff
@@ -53,7 +53,11 @@ let structure : type_definition -> type_structure = fun def ->
+          if !Clflags.typing_recovery
+          && Typing_recovery.erroneous_type_check type_expr
+          then Abstract 
+          else Synonym type_expr
```
Can you add a test for what's going on here?

trefis commented on typing/typemod.ml:
```diff
@@ -81,8 +81,31 @@ type error =
 
-exception Error of Location.t * Env.t * error
 exception Error_forward of Location.error
+exception Errors of Env.t * Typing_recovery.Error_set.t
```
I don't think `Errors` should take an environment, each of the individual errors already have the appropriate one.

xvw replied:
(already fixed by you, I guess?)

trefis commented on typing/typemod.ml:
```diff
@@ -81,8 +81,31 @@ type error =
+    if !Clflags.typing_recovery then
+      Typing_recovery.log_and_raise err
+    else
+      raise err
```
🙂

trefis commented on typing/typemod.ml:
```diff
@@ -129,7 +152,12 @@ let initial_env ~loc ~initially_opened_module
       Parse.simple_module_path lexbuf in
-        snd (type_open_ Override env loc {txt;loc})
+    try
+      snd (type_open_ Override env loc {txt;loc})
+    with
+    | Typetexp.Error.In_context _
+    | Env.Error.In_context _
+    | Persistent_env.Error _ when !Clflags.typing_recovery -> env
```
I'm not sure why we'd encounter an error from `Typetexp` here.
Also, merlin seems to catch (an equivalent of) `Cmi_format.Error`, and I guess you might want to do that.

trefis commented on typing/typemod.ml:
```diff
@@ -1587,325 +1614,339 @@ and transl_with ~loc env remove_aliases (rev_tconstraints, sg) constr =
-                  in
-                  Some (`Substituted_away subst)
+        try transl_sig_ env item srem
+        with Error.In_context _ when !Clflags.typing_recovery ->
```
If we want to be robust / extra safe I think we should adopt the same pattern as I suggested in `typecore`, i.e.
`try ... with exn when !Clags.typing_recovery && is_recoverable exn -> ...`

trefis commented on typing/typemod.ml:
```diff
@@ -1587,325 +1614,339 @@ and transl_with ~loc env remove_aliases (rev_tconstraints, sg) constr =
-                  Some (`Substituted_away subst)
+        try transl_sig_ env item srem
+        with Error.In_context _ when !Clflags.typing_recovery ->
+          transl_sig env srem
```
Also: that pattern is just not right.
You try to type check the signature once, if there's an error you drop the first element and try again ... even though the error might have been arbitrarily deep.
Basically you're dropping the whole prefix until the first error ... in a quadratic way 🙂 

`transl_sig_` should become `transl_sig_item` (and not receive srem!) and the rest of the fix should flow naturally.

trefis commented on typing/typemod.ml:
```diff
@@ -2355,13 +2402,34 @@ let simplify_app_summary app_view = match app_view.arg with
+       before starting and restore them when finished. *)
+    Typing_recovery_state.with_saved_types (fun () ->
+        try delayed ()
+        with Error.In_context _ when !Clflags.typing_recovery ->
```
```suggestion
        with exn when !Clflags.typing_recovery && is_recovarable exn ->
```

trefis commented on typing/typemod.ml:
```diff
@@ -2720,33 +2797,51 @@ and type_open_decl_aux ?used_slot ?toplevel ~funct_body names env od =
+            type_struct new_env shape_map srem
+          in
+          (str :: str_rem, sg @ sig_rem, shape_map, final_env)
+        with Error.In_context _ when !Clflags.typing_recovery ->
```
Same as everywhere else

trefis commented on typing/typemod.ml:
```diff
@@ -3289,7 +3397,19 @@ let type_implementation target initial_env ast =
         ignore @@ Warnings.parse_options false "-32-34-37-38-60";
       let (str, sg, names, shape, finalenv) =
-        type_structure initial_env ast in
+        (* Maybe can we find a better place to collect errors. *)
```
More importantly: `type_interface` doesn't get to `catch_errors` (and have them collected), so sadly no recovery in .mli files!

trefis commented on typing/typemod.ml:
```diff
@@ -3449,10 +3572,9 @@ let package_units initial_env objfiles target_cmi =
-      raise(Error(Location.in_file mli, Env.empty,
-                  Interface_not_compiled mli))
-    end;
+
```
(extra blank line here)

trefis reviewed (changes requested):
Still a partial review, but I thought it would be good to send this first batch of comments.

Additionally here are some things that are not in my review comments on the code, but which I don't want to forget:
- we should explicitly test for the absence of warnings in recovered contexts (it might be done already, but as a side effect of the extensive testing that this PR adds, without mentioning explicitly that some warnings do not trigger).
We can probably just add a `-w +a` in some or all of the new tests, and see what happens
- there is now a very regular handling of user facing errors (i.e. errors that actually get emitted, not internal ones) in most of the type checker ... but some of these units also define an `Error_forward` exception which ... are never recovered, they could be (by handling them just like non forward errors)
- ~the reporting of errors is still not acceptable (but I haven't looked at how it's implemented yet, I suspect it might be easily fixable)~ (it's now "ok")
- ~`Includemod` hasn't received the same treatment as other typechecker units, I'm not sure if this is intentional~ (discussed offline with Xavier, it is intentional et reasonable)

To sum up: there still are a few issues, but this PR is coming along quite nicely, and notably the code is now pretty regular and seems (to me) an improvement on the erratic way in which merlin was doing things.

trefis committed:
slightly improve report of errors with -typing-recovery

trefis commented:
I took the liberty of pushing a minimal improvement to error reporting, enough to make it acceptable.

xvw committed:
Remove `env` and `location` from `Env.Error.In_context`

trefis commented on typing/typing_recovery.mli:
```diff
@@ -0,0 +1,50 @@
+(**************************************************************************)
+
+(** Returns a reference initially set to false that will be set to true
+    when a type error is raised. *)
```
```suggestion
(** Returns a reference initially set to false that will be set to true
    when an error is logged (see below). *)
```

trefis commented on typing/typing_recovery.mli:
```diff
@@ -0,0 +1,50 @@
+val monitor_errors : unit -> bool ref
+
+val raise_error : exn -> unit
+[@@deprecated "use [log_and_raise] or [log_or_raise] instead"]
```
Should be removed.

trefis commented on typing/typing_recovery.mli:
```diff
@@ -0,0 +1,50 @@
+(* FIXME: should not mess with [Warnings.state] *)
+(** [catch_error ws errors_log f] gathers all the errors logged by [f ()] into
+    [errors_log]. *)
+val catch_errors : Warnings.state -> exn list ref -> (unit -> 'a) -> 'a
```
I still agree with the FIXME, I see no reason why `catch_errors` should take a `Warnings.state`

trefis commented on typing/typing_recovery.mli:
```diff
@@ -0,0 +1,50 @@
+
+
+val with_warning_attribute :
+  ?warning_attribute:Parsetree.attributes -> (unit -> 'b) -> 'b
```
Likewise: this shouldn't be here.

trefis reviewed (changes requested):
I finished reading through the changes to the typechecker and found no other issues (than the ones I reported yesterday).
I'm adding a few more comments on the "new" code, but it's about style and documentation at this point.

As for reading through the added tests: I ran out of steam. The tests exists, they might be redundant, not as precise as they could be, but they are present and plentiful.

xvw committed:
Break cyclic dependency

xvw committed:
Remove deprecated `raise_error`

xvw committed:
Remove redundant guard `typing_recovery`

xvw committed:
`Illegal_value_name` without recovery

xvw committed:
Recover on `Illegal_value_name`

xvw committed:
Add a test for `illegal_reference_to_recursive_class` in recovery

xvw committed:
Remove wrong comment about private constructor for exn

xvw committed:
Remove non-relevant `todo` comment

xvw committed:
Handle missing kind of errors in typing a pattern

xvw committed:
Document `uncatch_error` for `partial_pred`

xvw committed:
Fix wrong type parameter

xvw committed:
Track recoverable exception

xvw committed:
Some simplification in `typedecl`

xvw committed:
Add `Cmi_format.error` handling on initial_env

xvw committed:
Remove `Warning` mess

xvw committed:
Remove `with_warning_attribute` function

xvw committed:
Use `Typecore.is_recoverable` in `Typemod`

xvw committed:
diff test-suite

xvw committed:
Fix recovery in `modular explicit`

xvw committed:
Clarify comment for `ref_monitors_errors` double reference

xvw committed:
Make `transl_signature` not quadratic anymore

xvw committed:
Clarify handler in `transl_signature`

xvw committed:
Properly treat `Psig_extension` in `transl_signature`

xvw commented:
@trefis 

Thank you very much for the in-depth review!

I have tried to address each of the comments related to recovery, and I hope I have covered them all!

Next week, I will do a general review of the tests to address the comments that only related to the tests. Thank you!

trefis commented:
Thanks, I had a quick look through your latest changes, everything looks reasonable.
I'll do a proper review towards the end of next week.

xvw committed:
Reintroduce guard `Clflags.typing_recovery` for warnings

xvw committed:
Rebased against `trunk`

xvw committed:
Hygien

xvw committed:
Fix after rebasing from `trunk`

xvw committed:
Hygien

xvw committed:
Global pass on tests

xvw committed:
Hygien

trefis commented on typing/typemod.ml:
```diff
@@ -1587,325 +1614,315 @@ and transl_with ~loc env remove_aliases (rev_tconstraints, sg) constr =
+      ~save_part:(fun sg -> Cmt_format.Partial_signature sg)
+      (fun () ->
+         if keep_warnings then delayed ()
+         else Builtin_attributes.warning_scope [] delayed)
```
Unless you can provide a good justification for it (e.g. a test) let's remove that `keep_warnings` flag you introduced, simplify that code and push the `Builtin_attributes.warning_scope []` call inside `delayed`.

xvw replied:
Yes, seems remanent from Merlin, removed in https://github.com/ocaml/ocaml/pull/14241/commits/71802f53f88f36e058d31e5a389283bc8fc45b23

trefis commented on typing/typemod.ml:
```diff
@@ -2277,7 +2297,7 @@ let package_subtype env pack1 pack2 =
   in
   match mkmty pack1, mkmty pack2 with
-  | exception Error(_, _, Cannot_scrape_package_type r) ->
+  | exception Error.In_context (_, _, Cannot_scrape_package_type r) ->
       Result.Error (Errortrace.Package_cannot_scrape r)
```
I was worried about getting two errors in this case with `--typing-recovery` but that appears (on my tests) not to be the case.
I suppose this is (again) due to both errors sharing the same location. In any case, I get the same error message both with and without recovery, so this is fine.

(This comment is sort of a "note to self")

trefis commented on typing/typemod.ml:
```diff
@@ -2720,33 +2773,49 @@ and type_open_decl_aux ?used_slot ?toplevel ~funct_body names env od =
+          in
+          (str :: str_rem, sg @ sig_rem, shape_map, final_env)
+        with exn when !Clflags.typing_recovery && Typecore.is_recoverable exn ->
+          type_struct env shape_map srem
```
This has the same quadratic behaviour as your previous patch to `transl_signature`...

xvw replied:
I may be missing something, but it seems to me that the current implementation is similar to `transl_signature`? Here, each item is processed individually without recursively calling the `type_struct` function within `type_str_item`. However, I've adapted the handler in the same way as `transl_signature` here: https://github.com/ocaml/ocaml/pull/14241/commits/886dd1b8683db7399479595e09cb9ef8388cb581

WDYT?

trefis replied:
As discussed offline: I was indeed wrong.
But I do find the new version nicer. :)

trefis commented on typing/typemod.ml:
```diff
@@ -2720,33 +2773,49 @@ and type_open_decl_aux ?used_slot ?toplevel ~funct_body names env od =
 
-and type_structure ?(toplevel = false) ~funct_body anchor env sstr =
+and type_structure ?(toplevel = false)
+    ?(keep_warnings = false) ~funct_body anchor env sstr =
```
Same remark about `keep_warnings` as above: either justify or remove.

xvw replied:
Removed in https://github.com/ocaml/ocaml/pull/14241/commits/71802f53f88f36e058d31e5a389283bc8fc45b23

trefis commented on typing/typemod.ml:
```diff
@@ -3289,7 +3372,17 @@ let type_implementation target initial_env ast =
         ignore @@ Warnings.parse_options false "-32-34-37-38-60";
       let (str, sg, names, shape, finalenv) =
-        type_structure initial_env ast in
+        if !Clflags.typing_recovery then
+          let caught = ref [] in
+          let result =
+            Typing_recovery.catch_errors caught
+              (fun () -> type_structure initial_env ast)
+          in match !caught with
+          | [] -> result
+          | exn_list -> collect_recovery_errors target exn_list
+        else
+          type_structure initial_env ast
+      in
```
I expected to find a similar change in `type_interface` ... but didn't, which makes me wonder what happens for mli files.
While you added several `.ml` files to the testsuite, there's apparently no `.mli` in `testsuite/tests/typing-recovery`...

xvw replied:
Here: https://github.com/ocaml/ocaml/pull/14241/commits/476a622c6afcab345981a7fea452b0bc2cba60ac, I handle recovery in `type_interface` and in https://github.com/ocaml/ocaml/pull/14241/commits/203a2fcdfa36f6c8060c7de3c46accf14e351027 I add a test for `mli`.

trefis reviewed (changes requested):
Finally read through typemod, left some comments in problematic situations. The rest of that file is fine.

xvw committed:
Clarify Handler for `type_struct`

xvw committed:
Remove `keep_warning` (non-used flag)

xvw committed:
Handle recovery for interface files

xvw committed:
Test for illustrating `typing-recovery` in the case of interface

xvw committed:
Rebase on top of `trunk`

trefis commented on typing/typecore.ml:
```diff
@@ -4893,39 +5144,56 @@ and type_expect_
-        exp_attributes = sexp.pexp_attributes;
-        exp_env = env }
+      if !Clflags.typing_recovery then
+        let obj = type_exp env e in
```
I think this shouldn't be here: it's only useful in case of recovery (otherwise it's just redundant: `e` is typed as part of `suspended`).

Also: I wondered whether it made sense to type the object expression again in case of recovery (because "what if that is itself incorrect?"), I think it kind of does: we're recovering only in case the object doesn't have the method we are looking for (implying that the object expression itself was well typed).

I see some possible improvements here:
1. you move the line in question inside the handler, adding a comment explaining why it makes sense
2. a better improvement would be to change the type of `Undefined_method` to embed the object instead of just its type, so you don't have to recompute it
  ... and you also add a comment explaining what's going on

trefis commented on typing/typecore.ml:
```diff
@@ -5283,15 +5550,21 @@ and type_expect_
-        exp_desc = Texp_struct_item (si, exp);
-        exp_type = exp.exp_type;
-        exp_loc = loc;
-        exp_extra = [];
-        exp_attributes = sexp.pexp_attributes;
-        exp_env = env;
-      }
+      try delayed ()
+      with exn when !Clflags.typing_recovery && is_recoverable exn ->
+        (* the original error will already have been logged. *)
+        Typing_recovery.catch_errors (ref [])
```
Looking back at this, I think the comment could be a bit clearer, something like:
```
The original error has been logged already, and we don't want spurious errors to 
show up on parts that are recovered, so we locally redirect all errors to a ref that 
we never read from.
```

Or is that too much?

trefis replied:
Also: should there be a call to `erroneous_type_register` as in the other places where recovery happens?

trefis commented on typing/typedecl.ml:
```diff
@@ -1239,8 +1260,11 @@ let transl_type_decl env rec_flag sdecl_list =
-           let params = param_types tdecl.typ_params in
-           raise(Error(sdecl.ptype_loc, Unbound_type_var{ var; params; decl}))
+           if not (Typing_recovery.erroneous_type_check var) then
+             (* do not report spurious error, we're in a recovered context *)
```
```suggestion
           (* do not report spurious error in a recovered contexts *)
           if not (Typing_recovery.erroneous_type_check var) then
```

trefis commented on typing/typemod.ml:
```diff
@@ -2355,13 +2377,34 @@ let simplify_app_summary app_view = match app_view.arg with
+  if !Clflags.typing_recovery then
+    (* Recovery: when we start typing a module we don't want to
+       include potential saved_items from its parent. We backup them
+       before starting and restore them when finished. *)
+    Typing_recovery_state.with_saved_types (fun () ->
```
Is something special happening here that doesn't happen in the other places where we call `with_saved_types`?

(I'm trying to understand if the comment tries to warn me about something I'm currently not seeing, or if it's just noise)

xvw replied:
Nop, it was a relicat of the Merlin implem

trefis commented on typing/typemod.ml:
```diff
@@ -3275,6 +3343,20 @@ let gen_annot target annots =
     annots
 
+let collect_recovery_errors unit exn_list =
+  (* TODO: Improve multi error recovery *)
```
Let's remove that TODO.

trefis commented on typing/typemod.ml:
```diff
@@ -3720,13 +3822,21 @@ let report_error env ~loc err =
+    Typing_recovery.Error_set.to_list errors
+    |> List.concat_map (fun err -> err.Location.main :: err.Location.sub)
+  in
+  Location.errorf ~loc ~sub:errors "Multiple errors were encountered@,"
```
I think it would be nice in situations where there's only one error to not say : "multiple errors were encountered".

Currently the API doesn't let you do that but you could expose `Location.mkerror` (or wrote a `Location.error_of_msg`).

xvw replied:
Do you like the proposition in https://github.com/ocaml/ocaml/pull/14241/commits/043b2c67ee519824c092dcf6c84d023ef7ea2a5e

trefis replied:
It's great.

trefis reviewed (changes requested):
I looked over the whole diff to `typing/` once again (except for the new files), I didn't see any major issue, the remaining comments are about cleaning up the code and its documentation.

xvw committed:
Clarify comment (and register erroneous type) in typing items

xvw committed:
Remove remanent comments

xvw committed:
Change reporting of errors (when one error is recoverred)

xvw committed:
Update test-suite

xvw committed:
Embed `obj` inside `Undefined_method` exception

xvw committed:
Change recovery type

xvw committed:
Clarify `-typing-recovery` flag in manual

xvw committed:
Add section for `Typing recovery` in `typing/HACKING.adoc`

xvw committed:
Add sub-section for recovery (logging vs raising + testing)

xvw committed:
Clarify documentation

xvw committed:
Fix backward-misconception about user-facing logging for errors

trefis commented on typing/typing_recovery.mli:
```diff
@@ -0,0 +1,73 @@
+
+(** Logs the given exception (which will generally be caught by another
+    handler) if called from under a [catch_errors], and then raises
+    it. *)
```
```suggestion
(** [log_and_raise exn] raises [exn], but also logs it when called from under [catch_errors]. *)
```

trefis commented on typing/typing_recovery.mli:
```diff
@@ -0,0 +1,73 @@
+    and then synthesizes a consistent value, whereas [Typecore] will
+    interrupt type-checking to hand control over to [Typemod] (except
+    in certain cases, such as when processing function parameters or
+    inline records where one would like to continue recovery). *)
```
```suggestion
(** Logs the given exception if called from under a [catch_errors] otherwise raises it. *)
```

xvw committed:
Add missing `get_saved_types_from_attributes`

For Merlin instrumentation

trefis commented on testsuite/tests/typing-recovery/id_wrong_ty.compilers.reference:
```diff
@@ -0,0 +1,4 @@
+File "id_wrong_ty.ml", line 10, characters 28-38:
+10 | let rec id_wrong_ty : int = fun x -> x
+                                 ^^^^^^^^^^
+Error: This expression should not be a function, the expected type is int
```
AFAICT this doesn't test that recovery is happening?

trefis commented on testsuite/tests/typing-recovery/illegal_reference_to_recursive_class.compilers.reference:
```diff
@@ -0,0 +1,7 @@
+Error: This module type is recursive. This use of the recursive module Foo
+       within the definition of the module Bar
+       makes the module type of Bar depend on the module type of Foo.
+       Such recursive definitions of module types are not allowed.
```
Same for this test.

trefis commented on testsuite/tests/typing-recovery/unbound_type_constructor.compilers.reference:
```diff
@@ -0,0 +1,4 @@
+File "unbound_type_constructor.ml", line 12, characters 17-21:
+12 | let foo2 : int * toto = 3, 4
+                      ^^^^
+Error: Unbound type constructor toto
```
Idem.

trefis commented on typing/typing_recovery_state.mli:
```diff
@@ -0,0 +1,35 @@
+(**************************************************************************)
+
+(** Save fragments of the typedtree to CMT files so they can be processed
+    later by a server-side language (such as Merlin). *)
```
The fact that recovery reuses some types and storage from the cmt infrastructure is mostly historical / accidental.
The point never was to save the information in cmt as this comment seems to indicate.

A similar comment is added to HACKING.adoc and should IMO be removed. (We don't want to imply that cmts are needed for recovery or merlin)

trefis reviewed (commented):
And we're back to cosmetic / noisy remarks on my part!

I think we're basically done here.
You need to rebase one last time, at which point I'll *promptly* review one last time and then we're good to merge I think.

This PR history is very noisy and I don't think it's worth trying to cleanup so feel free to squash everything during your rebase.

trefis commented on Changes:
```diff
@@ -792,6 +792,12 @@ OCaml 5.5.0
+  than one typing error, facilitating integration with editors that
+  need to report more than one error.
+  (Xavier Van de Woestyne, Ulysse Gerard, Thomas Refis and Frédéric
+  Bour, review by Florian Angeletti and Thomas Refis)
```
@Octachron just a heads up that you're mentioned as a reviewer here, in case you want to edit yourself out.

(I know you've done some amount of reading and counseling on this PR, I don't want to minimize your work, I'm just not sure you want to be deemed responsible in case I let some bugs through)

Octachron replied:
Now that I have been successfully tricked into having a quick last look to the PR (except for the tests) this seems fine to me.

trefis reviewed (approved):
Good to go!

A quick recap of the work that has been done since the PR was first open:
- at least a trillion test have been added to ensure that the recovery works well on the core language (modules and classes are not tested as extensively)
- an API was defined and some static checks put in place to guide compiler devs in ensuring that recovery stays up to date as new errors and checks are added to the typechecker
- some documentation was written regarding that whole infrastructure, as well as how and when error recovery can happen

Overall this feels like a nice cleanup and clear improvement over the state of things in merlin.

Thanks Xavier!

trefis commented:
Btw, the CI failure seems spurious: there is some diff in the `parsing/` directory but it doesn't affect the parsetree (or any of the related datatypes), so I don't think ppxlib maintainers really need to have a look at this.
Nevertheless let's ping @pitag-ha (once upon a time ppxlib maintainer IIRC) and @gasche (who will know who to ping) as ppxlib _might_ version the `Location` API.

gasche commented:
(cc: maybe @NathanReb on the ppxlib front as well?)

Thanks @xvw and @trefis for all this demanding work. 

I would be curious to have a look to the developer-facing documentation to check that it is approachable by someone who is not already knowledgeable about Merlin. I will probably try to do this around now -- or find someone else motivated to do it -- but in principle this should be good to merge.

Octachron commented on typing/typemod.ml:
```diff
@@ -3324,6 +3369,19 @@ let gen_annot target annots =
+      ) exn_list (Typing_recovery.Error_set.empty)
+  in
+  raise (Errors (Location.in_file
+                   (Unit_info.input_source_file unit), error_list))
```
Maybe a comment indicating that in recovery mode, the location should point to the actual input file rather than the pretty source name for humans?

xvw replied:
Is that ok for you https://github.com/ocaml/ocaml/pull/14241/commits/e9f6d0358782a4979bbbf28ee5e5e0fee1e3cf8c ?

pitag-ha commented:
About ppxlib: Ppxlib does use part of the `Location` API. It deals with changes to `Location` via version-conditional compilation (i.e., a different way of handling versions than the versioning of the parsetrees). However, this PR just seems to add a new value to `Location` and I don't think that's of interest for Ppxlib. Still, just in case, it's good if the current Ppxlib maintainers @NathanReb and @patricoferris are aware of this small change :)

Octachron commented on man/ocamlc.1:
```diff
@@ -793,6 +793,11 @@ Force the left-hand part of each sequence to have type unit.
 .TP
+.B \-typing-recovery
+Experimental: let the compiler progress past the first type error,
+allowing it to additional type errors in the file. N.B. any error past
```
There is a missing verb here. "allowing it to detect"  or "allowing it to report" ? Also "additional type errors" might suggest that those type errors might refine the first one whereas the focus is probably more on unrelated type errors. Also "the file" is ambiguous in case like `ocamlc a.ml b.ml`. Maybe

> Experimental: let the compiler progress past the first type error,
in order to try to report multiple errors type errors at the same time.

?

xvw replied:
Done in https://github.com/ocaml/ocaml/pull/14241/commits/77e4efb86fc2487e095b959ead106902dea3d2d6

Octachron commented on typing/typecore.ml:
```diff
@@ -219,7 +219,159 @@ type error =
+    | Tpackage package ->
+        Tpackage (deep_copy_package copy package)
+    | Tfunctor (l, id, package, type_expr) ->
+        Tfunctor (l, id, deep_copy_package copy package, copy type_expr)
```
There may be an issue here, the unscoped idents should probably also be copied to freeze module-dependent paths.

trefis replied:
Indeed, I had missed that these idents could change. :see_no_evil: 
(For @xvw 's defense: I believe `Tfunctor` appeared following a rebase, and it's not too surprising that it fell through the cracks)

Octachron replied:
I completely agree that this is a fresh paint issue, and in fact I think the fix could be potentially left to a subsequent PR.

xvw replied:
(Should be) done in https://github.com/ocaml/ocaml/pull/14241/commits/40451600480026d799581a60003425bade205d1d

Octachron replied:
Copying the `M` ident in `fun (module M:S) -> M.t` is not enough, the old ident needs to be replaced by the copied version in all its occurrence. I propose to drop 4045160, and I will fix this part in a subsequent PR.

Octachron commented on testsuite/tests/typing-recovery/multiply_bound_variable.ml:
```diff
@@ -0,0 +1,21 @@
+let f = function
+  | (x, x, y) -> x + y
+  | _ -> 0
+
```
I think it would be a good idea to have a test which illustrates that this error is a `log_or_raise`. Maybe
```ocaml
let f = function
| Some x, Some x | [], x -> x
```
?
(And potentially the same for all `log_or_raise`.)

Octachron replied:
This part is still missing? My idea was to have one test that illustrate that the typechecking is going further after the `log_or_raise`, and thus we can detect type error after the non-linear binding.

Octachron commented on typing/typecore.ml:
```diff
@@ -219,7 +219,159 @@ type error =
+     pack_constraints =
+       List.map (fun (l, tl) -> l, copy tl) pack_constraints}
+
+  let deep_copy_desc copy = function
```
Also I think it would be nice to have a comment explaining what are the obstacles to using `Ctype.duplicate_type` or ones of the other copy function here.

trefis replied:
The difference in behavior with the other copying functions are small but important I think, though some deduplicating could happen by adding some flag to `Btype.copy_type_desc`.

The code added by the PR always wants a fresh copy of all the parts of the type regardless of their level, so:
- `Ctype.duplicate_type` is out: the type variables remain shared, we don't want that
- `Ctype.copy` is out too: it keeps the non generic parts of the type shared

I think a good middle ground would be to keep the `deep_copy` function added by this PR but have it called a slightly modified version of `Btype.copy_type_desc` (which would not assert false on `Tvariant` in particular).

But the code added by the PR is local and simple enough to have me believe that merging as is and leaving the cleanup to people more experienced with the low level bits of the typechecker is probably the best course of action.
Does that sound reasonable to you @Octachron ?

trefis replied:
(Though a comment could indeed be added in the code to explain that choice)

Octachron replied:
There is also `Subst.(typexp (for_saving identity))` which does copy type variables, but without keeping the level which will not work for error message.

Overall, I agree that having a local `copy/freeze` function is a reasonable step, but adding a comment to document the specification, and possible avenues of deduplication would be even better.

xvw replied:
I try to state that in https://github.com/ocaml/ocaml/pull/14241/commits/3e529a412b147f1cb96f25ca67d39cc33495e5b5, do you think it is ok?

Octachron replied:
I agree that this is a nice documentation comment.

Octachron commented:
@gasche , the developer documentation looks fine to me. Overall, I find the current state very nice since outside of recovery points, the main change is to use `log_and_raise` rather than raising an error, and optionally use `log_or_raise` if it is somehow possible to ignore the raised error at least during typechecking.

gasche commented:
Great! Thanks for having a look, I'll let you make merge decisions.

NathanReb commented:
From a quick look it seems nothing here will cause trouble for ppxlib! Thanks for the heads up though, don't hesitate to ping us on other such PRs!

trefis commented:
Ping @Octachron, this PR already needs a new rebase (because of conflicts with one of your PRs that got merged into trunk). More generally: this PR is somewhat sensitive to changes in the typer (they won't necessarily show up as conflict, but the build would fail post merge).

So now that we're close to being done and after having spread my review over a few months I  feel compelled to get it merged quickly if possible (and avoid more work for @xvw).

Octachron commented:
@trefis, this is ready to merge from my point of view, the bug with respect to unscoped idents can be fixed later.

trefis commented:
Based on @Octachron last comment I took the liberty to merge.

Thanks everyone!

---
PS: I also noticed a typo (that can be fixed in a follow up PR) in the manpages after giving one last look before merging:
> to report multiple errors type errors at the same time.
The word "errors" is present one too many times.

trefis commented:
And another blip: the change entry ended up being in the 5.5 section.
@Octachron since I'm apparently pushing work on your stack today: I'll let you decide if this change should be backported to 5.5 or if the changelog entry should be moved. :pray:

Octachron commented:
I was planning to do some Changes cleanup anyway, and I will add the typo fix to my current manpage PR.

Thanks for the PR @xvw , and thanks for the review @trefis.

xvw commented:
Sorry for the minor glitches at the end of the process :)
A huge thank you to @trefis for his help, guidance, and reviews, and to @octachron for his eagle eye and the speed with which he conducted the review. I’m thrilled with this collaboration! Thank you!
